File: ios/background_beacon_sdk/Sources/background_beacon_sdk/BeaconModels.swift
import CoreLocation

// Models
struct BeaconRegionData {
    let identifier: String
    let uuid: UUID
    let major: CLBeaconMajorValue?
    let minor: CLBeaconMinorValue?

    init?(map: [String: Any?]) {
        guard let identifier = map["identifier"] as? String,
              let uuidString = map["uuid"] as? String,
              let uuid = UUID(uuidString: uuidString)
        else { return nil }
        self.identifier = identifier
        self.uuid = uuid
        self.major = (map["major"] as? NSNumber).map { CLBeaconMajorValue(truncating: $0) }
        self.minor = (map["minor"] as? NSNumber).map { CLBeaconMinorValue(truncating: $0) }
    }

    /// Region for monitoring
    var clRegion: CLBeaconRegion {
        let region: CLBeaconRegion
        if let major, let minor {
            region = CLBeaconRegion(uuid: uuid, major: major, minor: minor, identifier: identifier)
        } else if let major {
            region = CLBeaconRegion(uuid: uuid, major: major, identifier: identifier)
        } else {
            region = CLBeaconRegion(uuid: uuid, identifier: identifier)
        }
        region.notifyEntryStateOnDisplay = true
        return region
    }

    var constraint: CLBeaconIdentityConstraint {
        if let major, let minor {
            return CLBeaconIdentityConstraint(uuid: uuid, major: major, minor: minor)
        }
        if let major {
            return CLBeaconIdentityConstraint(uuid: uuid, major: major)
        }
        return CLBeaconIdentityConstraint(uuid: uuid)
    }

}

struct ScanSettingsData {
    /// duration/notification are Android concerns
    /// (iOS region monitoring is scheduled by the OS itself)
    let rangingEnabled: Bool

    /// Location keep-alive so ranging keeps flowing in the background —
    /// session-scoped: starts on first region enter, stops after leaving
    /// all regions (conditions/costs in the ScanSettings dartdoc).
    let continuousRanging: Bool

    /// Ranged-event flush cadence — matches reportDelay on Android
    /// (CL's didRange fires ~1/s, too often for the "one event per cycle"
    /// contract).
    let scanIntervalMs: Int

    init?(map: [String: Any?]) {
        guard let rangingEnabled = map["rangingEnabled"] as? Bool else { return nil }
        self.rangingEnabled = rangingEnabled
        self.continuousRanging = (map["continuousRanging"] as? Bool) ?? false
        // Dart always sends it (required) — fallback just keeps the struct non-nil
        self.scanIntervalMs = (map["scanIntervalMs"] as? Int) ?? 5000
    }
}

/// CLBeacon → Map per the Dart `Beacon` wire contract.
/// txPower: CoreLocation doesn't expose it → -1 / distance: `accuracy`
/// in meters (-1 = unknown). `mac` is deliberately omitted (iOS never
/// exposes MAC) — Dart reads the missing key as null.
func beaconMap(_ beacon: CLBeacon) -> [String: Any] {
    [
        "uuid": beacon.uuid.uuidString.lowercased(),
        "major": beacon.major.intValue,
        "minor": beacon.minor.intValue,
        "rssi": beacon.rssi,
        "txPower": -1,
        "distance": beacon.accuracy,
        "lastSeen": iso8601(Date()),
    ]
}

/// Map per the Dart `BeaconEvent` wire contract.
/// type must match the enum names exactly: "enterRegion" | "exitRegion" | "ranged"
/// ("monitoringPaused" | "monitoringResumed" exist in the contract but are
/// Android-only — iOS never emits them)
func eventMap(type: String, region: String, beacons: [[String: Any]]) -> [String: Any] {
    [
        "type": type,
        "region": region,
        "beacons": beacons,
        "timestamp": iso8601(Date()),
    ]
}

// Local time without timezone suffix — matches DateTime.parse on Dart and iso8601 on Kotlin
private let iso8601Formatter: DateFormatter = {
    let formatter = DateFormatter()
    formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS"
    // POSIX locale guards against device-locale drift (Buddhist calendar / 12-hour)
    formatter.locale = Locale(identifier: "en_US_POSIX")
    return formatter
}()

func iso8601(_ date: Date) -> String {
    iso8601Formatter.string(from: date)
}

File: android/src/main/kotlin/com/example/background_beacon_sdk/core/BeaconModels.kt
package com.example.background_beacon_sdk.core

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

// Kotlin data classes mirroring the Dart models.
// For the wire contract (key names / formats) the dartdoc in lib/src/models/ is the spec.

data class BeaconRegionData(
    val identifier: String,
    /** null = wildcard matching every UUID (the chip filter already catches all iBeacons) */
    val uuid: String?,
    val major: Int?,
    val minor: Int?,
) {
    companion object {
        @Suppress("UNCHECKED_CAST")
        fun fromMap(map: Map<String, Any?>): BeaconRegionData = BeaconRegionData(
            identifier = map["identifier"] as String,
            // Dart already normalizes to lowercase — lowercasing again guards one-sided slips
            uuid = (map["uuid"] as String?)?.lowercase(),
            major = (map["major"] as Number?)?.toInt(),
            minor = (map["minor"] as Number?)?.toInt(),
        )
    }

    /// null uuid/major/minor = wildcard matching any value
    fun matches(uuid: String, major: Int, minor: Int): Boolean =
        (this.uuid == null || this.uuid == uuid) &&
            (this.major == null || this.major == major) &&
            (this.minor == null || this.minor == minor)
}

data class ScanSettingsData(
    val scanIntervalMs: Int,
    val scanDurationMs: Int,
    val foregroundServiceNotification: Boolean,
    val beaconLayout: String,
    val rangingEnabled: Boolean,
    /** iOS-only (location keep-alive) — Android accepts it per contract but never uses it */
    val continuousRanging: Boolean = false,
) {
    /**
     * How long a region stays silent before it counts as exited — must cover
     * at least 2 full missed scan cycles (a beacon can drop one cycle on a
     * signal dip) and never below 10 s so short intervals can't make exit
     * flap. Used by both in-process (BleBeaconScanner) and headless
     * (HeadlessBeaconRunner).
     */
    val exitTimeoutMs: Long
        get() = maxOf(10_000L, scanIntervalMs * 2L + scanDurationMs)

    companion object {
        fun fromMap(map: Map<String, Any?>): ScanSettingsData = ScanSettingsData(
            scanIntervalMs = (map["scanIntervalMs"] as Number).toInt(),
            scanDurationMs = (map["scanDurationMs"] as Number).toInt(),
            foregroundServiceNotification = map["foregroundServiceNotification"] as Boolean,
            beaconLayout = map["beaconLayout"] as String,
            rangingEnabled = map["rangingEnabled"] as Boolean,
            // safe-cast: key added later — older persisted JSON lacks it
            continuousRanging = (map["continuousRanging"] as? Boolean) ?: false,
        )
    }
}

data class BeaconData(
    val uuid: String,
    val major: Int,
    val minor: Int,
    val rssi: Int,
    val txPower: Int,
    val distance: Double,
    val lastSeen: Date,
    /** The device's Bluetooth MAC — debug/inventory only (absent on iOS) */
    val mac: String? = null,
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "uuid" to uuid,
        "major" to major,
        "minor" to minor,
        "rssi" to rssi,
        "txPower" to txPower,
        "distance" to distance,
        "lastSeen" to iso8601(lastSeen),
        "mac" to mac,
    )
}

data class BeaconEventData(
    // Must match the Dart enum names exactly: "enterRegion" | "exitRegion" | "ranged"
    val type: String,
    val region: String,
    val beacons: List<BeaconData>,
    val timestamp: Date = Date(),
) {
    fun toMap(): Map<String, Any> = mapOf(
        "type" to type,
        "region" to region,
        "beacons" to beacons.map { it.toMap() },
        "timestamp" to iso8601(timestamp),
    )
}

// Local time without timezone suffix — matches DateTime.parse on Dart.
// SimpleDateFormat is not thread-safe, so create per call (event volume is low enough).
internal fun iso8601(date: Date): String =
    SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS", Locale.US).format(date)
